game-core/src/main/java/games/strategy/triplea/ai/pro/AbstractProAi.java

```diff
@@ -0,0 +1,513 @@
+import games.strategy.triplea.attachments.PoliticalActionAttachment;
+import games.strategy.triplea.delegate.DelegateFinder;
+import games.strategy.triplea.delegate.DiceRoll;
+import games.strategy.triplea.delegate.Matches;
+import games.strategy.triplea.delegate.PoliticsDelegate;
+import games.strategy.triplea.delegate.battle.BattleDelegate;
+import games.strategy.triplea.delegate.battle.IBattle;
+import games.strategy.triplea.delegate.battle.IBattle.BattleType;
+import games.strategy.triplea.delegate.data.CasualtyDetails;
+import games.strategy.triplea.delegate.data.CasualtyList;
+import games.strategy.triplea.delegate.remote.IAbstractPlaceDelegate;
+import games.strategy.triplea.delegate.remote.IMoveDelegate;
+import games.strategy.triplea.delegate.remote.IPurchaseDelegate;
+import games.strategy.triplea.delegate.remote.ITechDelegate;
+import games.strategy.triplea.odds.calculator.IBattleCalculator;
+import games.strategy.triplea.ui.TripleAFrame;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.logging.Level;
+import lombok.Getter;
+import org.triplea.java.collections.CollectionUtils;
+import org.triplea.util.Tuple;
+
+/** Pro AI. */
+public abstract class AbstractProAi extends AbstractAi {
+
+  private final ProOddsCalculator calc;
+  @Getter private final ProData proData;
+
+  // Phases
+  private final ProCombatMoveAi combatMoveAi;
+  private final ProNonCombatMoveAi nonCombatMoveAi;
+  private final ProPurchaseAi purchaseAi;
+  private final ProRetreatAi retreatAi;
+  private final ProScrambleAi scrambleAi;
+  private final ProPoliticsAi politicsAi;
+
+  // Data shared across phases
+  private Map<Territory, ProTerritory> storedCombatMoveMap;
+  private Map<Territory, ProTerritory> storedFactoryMoveMap;
+  private Map<Territory, ProPurchaseTerritory> storedPurchaseTerritories;
+  private List<PoliticalActionAttachment> storedPoliticalActions;
+  private List<Territory> storedStrafingTerritories;
+
+  public AbstractProAi(
+      final String name, final IBattleCalculator battleCalculator, final ProData proData) {
+    super(name);
+    this.proData = proData;
+    calc = new ProOddsCalculator(battleCalculator);
+    combatMoveAi = new ProCombatMoveAi(this);
+    nonCombatMoveAi = new ProNonCombatMoveAi(this);
+    purchaseAi = new ProPurchaseAi(this);
+    retreatAi = new ProRetreatAi(this);
+    scrambleAi = new ProScrambleAi(this);
+    politicsAi = new ProPoliticsAi(this);
+    storedCombatMoveMap = null;
+    storedFactoryMoveMap = null;
+    storedPurchaseTerritories = null;
+    storedPoliticalActions = null;
+    storedStrafingTerritories = new ArrayList<>();
+  }
+
+  @Override
+  public void stopGame() {
+    super.stopGame(); // absolutely MUST call super.stopGame() first
+    calc.stop();
+  }
+
+  @Override
+  public PlayerType getPlayerType() {
+    return PlayerType.PRO_AI;
```

Comment:
Should this remain as an abstract method?
At first glance it seems like Fast AI would want to return its own player type.

Reply:
Right, wanted to make this change, but forgot to do so